ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PmtTransactionPool.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright contributors to Hyperledger Besu.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.ethereum.privacy;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.BlockAddedEvent;
+import org.hyperledger.besu.ethereum.chain.BlockAddedObserver;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PmtTransactionPool implements BlockAddedObserver {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Map<Hash, PmtTransactionTracker> pmtPool;
+
+  public PmtTransactionPool() {
+    this.pmtPool = new HashMap<>();
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    // remove from the map any transactions that went into this block
+    // TODO take into account re-orgs
+    event
+        .getAddedTransactions()
+        .forEach(
+            tx -> {
+              LOG.debug("removing " + pmtPool.containsKey(tx.getHash()));
+              pmtPool.remove(tx.getHash());
+            });
+  }
+
+  public Optional<Long> getMaxMatchingNonce(final String sender, final String privacyGroupId) {
+
+    return pmtPool.values().stream()
+        .filter(
+            pmt ->
```

Comment:
these are the trackers, not pmts. This is a bit misleading ...

Reply:
done